packages/firestore/src/model/document.ts

```diff
@@ -24,55 +24,162 @@ import { FieldPath } from './path';
 
 import * as api from '../protos/firestore_proto_api';
 
+export interface MissingDocumentOptions {
+  hasCommittedMutations?: boolean;
+}
+
 export interface DocumentOptions {
   hasLocalMutations?: boolean;
   hasCommittedMutations?: boolean;
+
+  /**
+   * Memoized serialized form of the document for optimization purposes (avoids
+   * repeated serialization). Might be undefined.
+   */
+  memoizedProto?: api.Document;
 }
 
 /**
- * The result of a lookup for a given path may be an existing document or a
- * marker that this document does not exist at a given version.
+ * At a high level, Documents either exist or don't exist. However, when they
+ * don't exist there are different reasons why they don't exist and we treat
+ * these all differently in the cache.
  */
-export abstract class MaybeDocument {
-  constructor(readonly key: DocumentKey, readonly version: SnapshotVersion) {}
+export enum DocumentType {
```

Comment:
I would put MISSNG / EXISTS at the top of this enum, since they are the simple cases...